### PR TITLE
Fixed parsing of TLS messages that span multiple records

### PR DESCRIPTION
**Pull request by chtsanti**
... if the servers contain huge certificates, eg while connecting to
1000-sans.badssl.com.  In this cases the server handshake 
messages (which include server certificates) may sent to more than
one TLS records. If squid reads these TLS records in one read, it
fails to parse them correctly and the connection freezes in
squid side waiting for more server TLS data.
    
This is a Measurement Factory project

### CODE REVIEW

**Comment by chtsanti** [comment]
> Unfortunately, I cannot quickly tell where the bug was, and, hence, I cannot offer any specific fix suggestions. To move forward, can you explain why the old code does not work?

Imagine the case

- a TLS hello message is split and transferred using 2 TLS records.
- Also the TLS message receiver received both of the records 

The old code does the following:
The hello receiver calls the Security::HandshakeParser::parseHello to parse the  hello message.So the following sequence is executed:
 
1. call Security::HandshakeParser::parseRecord to parse a TLS record
2. calls the HandshakeParser::parseModernRecord(); This is parses one record add its content to HandshakeParser::tkMessages and calls the HandshakeParser::parseMessages.
3. The HandshakeParser::parseMessages throws an exception because the message exist in first record is incomplete.
4. The HandshakeParser::parseHello catch the exception and return  "it needs more data"/false

I must note that there are more records in data we are trying to parse, so the "needs more data" is not true. On the next step Squid schedules a new read from server to read more data but if we have read all hello, more data are never comes.

The new code try to do the following inside HandshakeParser::parseModernRecords:
 1. Parse a record and add its data ti HandshakeParser::tkMessages. If not enough data throw.
 2. If at least one record is parsed peek at the next record. If it has the same data type and we have all of the next record, parse it and add its data to HandshakeParser::tkMessages.
 3. Repeat the last step until all records we have and have the same type are  parsed.
 4. Parse messages in HandshakeParser::tkMessages

I must note that we should not throw if we have read at least one record, this is why I add the TLSPlainTextPeeker

**Comment by rousskov** [comment]
I see where the bug is now, thank you.

> 1. call Security::HandshakeParser::parseRecord to parse a TLS record

OK.
 
>  2. calls the HandshakeParser::parseModernRecord(); This is parses one record add its content to HandshakeParser::tkMessages 

OK.

> and calls the HandshakeParser::parseMessages.

This is where we need to make the decision whether to keep the current "concurrent" parse-messages-while-parsing-records design or switch to a new "sequential" accumulate-same-type-records-then-parse-their-fragments design where we accumulate fragments of the same type and then call parseMessages(). Both designs can be fixed/implemented correctly, but after sketching the solution for the former, I think it would be easier to implement the latter instead.

The key in the sequential design is that we cannot accumulate fragments from _all_ same-type records. Without parsing embedded  messages, the record parser itself cannot tell when to stop waiting for more records of the last type. For everything to work, we can only accumulate same-type records that we have read already. If that accumulation is not enough to parse the embedded messages, then throwing InsufficientInput is exactly what the algorithm should do.

I may be missing something, but I think the above accumulation is possible without peaking at the next record. How about calling parseMessages() from exactly two places:

- When tkRecords is atEnd() in the very beginning of parseModernRecord(), _before_ we attempt to parse the next record. `tkMessages.reinput(fragments, true)`. expectMore is true here because more same-type records may come.
- When the `currentContentType != record.type` is true in parseModernRecord(), _before_ we reset fragments to the newly parsed fragment. `tkMessages.reinput(fragments, false)`. expectMore is false here because I assume that a single message cannot be split across non-sequential records.

To reduce new code duplication, you should probably adjust parseMessages() to take a boolean expectMore parameter and start parseMessages() with a `tkMessages.reinput(fragments, expectMore)` call, before calling tkMessages.rollback() and entering the current `for` loop.

Please note that when currentContentType is the same as record.type in parseModernRecord(), we still append fragments but no longer adjust tkMessages because there is no need to adjust tkMessages when we are not calling parseMessages().

No peeking at the next record.

Any flaws in the above scheme that you cannot address?

**Comment by rousskov** [comment]
> When tkRecords is atEnd() in the very beginning of parseModernRecord(), before we attempt to parse the next record ...

Actually, it would be better to do that at the very _end_ of parseModernRecord(). There, there will be no questions whether `currentContentType` value is usable. There is already a parseMessages() call at the end of parseModernRecord(). After the fix we are discussing, that call becomes conditional on  tkRecords being atEnd().

**Comment by chtsanti** [comment]
I build a new branch  the SQUID-457-Unexpected-ERR_CANNOT_FORWARD-alternate which implements your proposal. It works well.

The only problem I am seeing is that if the TLS server records are partially received (eg because the connection is closed while received) then the HandshakeParser may not parse all of the received messages it can parse because the tkRecords.atEnd() will not be true.
The old code does not have this problem, and this is was one of the problems I was trying to avoid.

This is maybe is not a huge problem, because the connections failed, but in the other hand even some partially received info (eg server TLS release and some TLS extension features) maybe is useful when log these (closed) connections.  

OK this is can be overcome by force calling the parseMessages on partial received, but is n't it a more clear solution to peek at the next record and if you have it include it in the data you are going to parse, if not try to parse what you already have.

**Comment by rousskov** [comment]
> The only problem I am seeing is that if the TLS server records are partially received then the HandshakeParser may not parse all of the received messages it can parse because the tkRecords.atEnd() will not be true.

I assume that you mean "will not _immediately_ parse some of the received messages (while tkRecords.atEnd() remains false)". In other words, it will parse them _eventually_, but not right away.

Yes, there is a natural trade-off here between parsing (incomplete) messages multiple times and learning new information sooner. The new "sequential" accumulate-same-type-records-then-parse-their-fragments design delays parsing until the next record or the end of input, optimizing CPU usage. I do not think we want to resolve this trade-off the other way: spending CPU cycles on learning new information ASAP (as long as we stop parsing everything without any extra delays or pauses) but please correct me if I am wrong. 

> The old code does not have this problem, and this is was one of the problems I was trying to avoid.

Why you were trying to avoid delaying when messages are parsed if that delay can save us some CPU cycles? Given two correct implementations, the end result and the timing of that end result delivery would be the same, right?

> is n't it a more clear solution to peek at the next record

IMO, custom peeking (_as it was implemented_) opens a Pandora box: Without peeking, the parser is always straightforward, single-pass algorithm -- parse if you have enough data or throw if you need more. Custom peeking adds a new, completely different mode -- carefully interpret the data you already have while avoiding calling the existing parsing code to parse that data (because that code may throw). I am surprised you do not see how that peeking implementation already duplicates a lot of complex parsing logic. You are essentially adding another/alternative record parser (and the one that cannot use the protections offered by the existing tokenizer). And it can only get worse with time.

I believe we could "peek" without opening that Pandora box by adding something like a parseRecordMetadata parser and splitting record metadata parsing from record payload parsing. The correct implementation of that would not be really peeking in a classical sense -- we would still use a straightforward, single-pass algorithm for everything. I do not know if that implementation can be done without making parsing code look more complex/fragmented, but I suspect it would be difficult because to avoid reparsing, one would have to save parsed record metadata somewhere between parsing steps. We do not need to save/maintain that metadata now. Said that, if you insist on peeking, you can try this alternative.

**Comment by chtsanti** [comment]
> I assume that you mean "will not _immediately_ parse some of the received messages (while tkRecords.atEnd() remains false)". In other words, it will parse them _eventually_, but not right away.

Not exactly. I am refer to the following case:
  - The squid sent TLS hello
  - The TLS server start sending its hello and other handshake messages using A records.
  - The A-1 records are of the same type.
  - After the A-1 records and some bytes from the last record received by squid the connection aborted for a reason (eg the server closed the connection.)

Even if the A-1 records has enough information to be parsed from squid the code will not do it. OK it is an abnormally terminated connection so maybe we do not care about. However logging the received info (TLS details), even for an aborted connection maybe worths it. 

> The new "sequential" accumulate-same-type-records-then-parse-their-fragments design delays parsing until the next record or the end of input, optimizing CPU usage. 

The CPU usage is enough and good argument. But in the other hand I am not expecting many hello or certificate TLS messages which are not fit in a TLS record. However the server for its own reasons may select to sent small TLS records.

> I believe we could "peek" without opening that Pandora box by adding something like a parseRecordMetadata parser and splitting record metadata parsing from record payload parsing. 

This is not a fully solution because still it will not cover the case not full record metadata read from server. However maybe this is  will allow to cover more cases.

> The correct implementation of that would not be really peeking in a classical sense -- we would still use a straightforward, single-pass algorithm for everything. I do not know if that implementation can be done without making parsing code look more complex/fragmented

I am sure it can not :-)

> I am surprised you do not see how that peeking implementation already duplicates a lot of complex parsing logic. You are essentially adding another/alternative record parser (and the one that cannot use the protections offered by the existing tokenizer). 

Of course I saw it. Initially I tried to add peeking functionality inside BinaryTokenizer, and then just use a flag inside existing TLSPlaintext, but with no very good results. So I used a new TLSPlainTextPeeker (which finally can be merged to TLSPlainText).

I am not insist that the straightforward, single-pass algorithm is beter, I just have the above problem to solve. And a Peeker class looked as the simpler method which does not add a huge complexity.

So I think we have the following options:
 1) We do not care about aborted transactions, we are not expecting many of them. In this case the SQUID-457-Unexpected-ERR_CANNOT_FORWARD-alternate branch is ok.
 2) Split TLSPlaintext into TLSPlainTextMedata and TLSPlainTextData however still we have to add checks if we can read the Record, the record type etc. The code will not look very simple.
 3) I supose you are not a fan of a solution where we could merge TLSPlainTextPeeker and TLSPlaintext. Definitely will not produce a simple straightforward code the TLSPlainText would need peek and doParse methods.
 4) Maybe just force a Handshake::parseMessages when the connection aborted before all of the server handshake received? But this is requires action inside PeerConnector, I do not like.

I do not believe the (1) will cause huge problems, we are not expecting many of such cases. Finally we can log something like "partially received handshake", with no details.

**Comment by rousskov** [comment]
```C++
fragment = tk.pstring16(".fragment");
```

Assumption: The above line in the official TLSPlaintext constructor means that the official parser will not look at partially-received records. Please correct me if I am wrong.


> 4. force a Handshake::parseMessages when the connection aborted 

Yes, "flushing" a buffering parser like that would be the overall best solution to interpreting partial TLS input IMO. Is it necessary? I doubt it is -- the current parser does not flush AFAICT (see above). I think this functionality can be left unimplemented.

I understand that the official code would parse N-1 fragments from the full records before stumbling on the last partial record N. The new PR code will not parse any of those fragments. We _are_ reducing detail coverage, but I think that reduction is insignificant and is worth keeping the parser simple.

Do we want to detail the reason why a transaction failed? Yes. Do we have to parse partial TLS records to do that? In most cases, no. The transaction failed not because there was something special/incompatible in those partial TLS records. The transaction failed because the server aborted the TCP connection. And we ought to be able to detail _that_ without parsing those partial TLS records.

In other words, it is a TCP-level error, not a TLS-level error.

There is a small area where the two errors overlap: The server thought it sent everything, but Squid thought that the record is still missing some fragment bytes. Ideally, we should add the latter "while waiting for the remainder of the TLS record" context detail to the TCP-level error message, but we do not have to do that in this PR IMO.

Please note that the admin can (or should be able to) log handshake bytes (including partial handshake bytes) to access.log if they want to triage using TLS-level details.

**Comment by chtsanti** [comment]
We can close this PR.
I create the PR 46  which implements Alex suggestion about this issue.

**Comment by rousskov** [comment]
Closing. PR #46 contains the latest version of this work now.
